Source/WebCore/platform/ios/VideoFullscreenInterfaceAVKit.mm

```diff
@@ -98,6 +104,296 @@ @interface AVPlayerViewController (Details)
+}
+@property (assign) VideoFullscreenInterfaceAVKit* fullscreenInterface;
+@end
+
+@implementation WebAVPictureInPictureController
+
+- (VideoFullscreenInterfaceAVKit*)fullscreenInterface
+{
+    ASSERT(isMainThread());
+    return _fullscreenInterface.get();
+}
+
+- (void)setFullscreenInterface:(VideoFullscreenInterfaceAVKit*)fullscreenInterface
+{
+    ASSERT(isMainThread());
+    _fullscreenInterface = *fullscreenInterface;
+}
+
+- (instancetype)initWithFullscreenInterface:(VideoFullscreenInterfaceAVKit *)interface
+{
+    if (!(self = [super init]))
+        return nil;
+
+    [self setFullscreenInterface:interface];
+
+    auto* playerController = static_cast<AVPlayerController*>(interface->playerController());
+    _controller = adoptNS([allocWebAVPictureInPictureContentViewControllerInstance() init]);
+    [_controller setController:playerController];
+    LOG(Fullscreen, "initWithFullscreenInterface isPlaying:%d", [playerController isPlaying]);
+
+    auto source = adoptNS([allocAVPictureInPictureControllerContentSourceInstance() initWithSourceView:static_cast<UIView*>(interface->playerLayerView()) contentViewController:_controller.get() playerController:playerController]);
+
+    _pip = adoptNS([allocAVPictureInPictureControllerInstance() initWithContentSource:source.get()]);
+    [_pip setDelegate:self];
+
+    return self;
+}
+
+- (void)dealloc
+{
+    if (_startPictureInPictureTimer)
+        [self removeObserver];
+    [_startPictureInPictureTimer invalidate];
+    _startPictureInPictureTimer = nil;
+    _controller = nil;
+    _pip = nil;
+    [super dealloc];
+}
+
+- (void)tryToStartPictureInPicture
+{
+    if (_startPictureInPictureTimer)
+        return;
+
+    if (_pip.get().pictureInPicturePossible) {
+        dispatch_async(dispatch_get_main_queue(), ^{
+            [self->_pip startPictureInPicture];
+        });
+        return;
+    }
+
+    _startPictureInPictureTimer = [NSTimer scheduledTimerWithTimeInterval:startPictureInPictureTimeInterval repeats:NO block:^(NSTimer *_Nonnull) {
+        [self removeObserver];
+        self->_startPictureInPictureTimer = nil;
+        if (_fullscreenInterface)
+            _fullscreenInterface->failedToStartPictureInPicture();
+    }];
+
+    [self initObserver];
+}
+
+- (void)stopPictureInPicture
+{
+    [_startPictureInPictureTimer invalidate];
+    _startPictureInPictureTimer = nil;
+    [self removeObserver];
+    [_pip stopPictureInPicture];
+}
+
+- (void)cleanup
+{
+    [_pip setDelegate:nil];
+    [self stopPictureInPicture];
+}
+
+- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
+{
+    if (!_startPictureInPictureTimer)
+        return;
+    [_startPictureInPictureTimer invalidate];
+    _startPictureInPictureTimer = nil;
+
+    [self removeObserver];
+
+    dispatch_async(dispatch_get_main_queue(), ^{
+        [self->_pip startPictureInPicture];
+    });
+}
```

Comment:
I don't get this at all :)

Why do you start PiP when someone observes any key path?